### DISCUSSION TITLE
Why Http headers is being vanished when using ORJSONResponse() in fastAPI?

### DISCUSSION THREAD

**Discussion by ahadrt** [fastapi/fastapi#7177] [Questions] [answered]
Http headers can be modified in [fastAPI](https://fastapi.tiangolo.com/); however when I tried to use faster serializer (in this instance `ORJSONResponse`) in order to have faster response, all added or modified headers confusingly is being vanished. **I don't have any idea that why it happens and how can I fix it?**
also we got the same result for `UJSONResponse` too.

*thanks for sharing your experiences.* 

### A. This code can't modify headers
```
from fastapi import APIRouter, Response
from fastapi.responses import ORJSONResponse
router = APIRouter()

@router.get("/test")
async def plan_list(response:Response):
    response.headers["test-header"]="test-value"
    return ORJSONResponse(list(range(10000)))
```

### Output header list of code A:
```
date: Thu, 22 Oct 2020 18:42:19 GMT
server: uvicorn
content-length: 48891
content-type: application/json
```

### B. This code modifies headers normally 
```
from fastapi import APIRouter, Response
from fastapi.responses import ORJSONResponse
router = APIRouter()

@router.get("/test")
async def plan_list(response:Response):
    response.headers["test-header"]="test-value"
    return list(range(10000))
```
### Output header list of code B:
```
date: Thu, 22 Oct 2020 18:42:19 GMT
server: uvicorn
content-length: 48891
content-type: application/json
test-header: test-value
```

test platform:
* OS: macOS 10.15.7
* client: insomnia 2020.14.1
* python: 3.8.5

**Comment by includeamin** [comment]
try this
``` python
from fastapi import APIRouter, Response
from fastapi.responses import ORJSONResponse

router = APIRouter()


@router.get("/test", response_class=ORJSONResponse)
async def plan_list(response: Response):
    response.headers["test-header"] = "test-value"
    return list(range(10000))

```

**Comment by ahadrt** [comment]
@includeamin  Actually, as I [tested](https://stackoverflow.com/questions/64488653/why-http-headers-is-being-vanished-when-using-orjsonresponse-in-fastapi) before, using `response_class=ORJSONResponse` is more than 10 times slower than calling the `ORJSONResponse()`directly. Specifically when data is very large object >10 MB I have to use it as function call.

**Comment by includeamin** [comment, marked as answer]
I'm not sure that it's good to do that or not.
if you want to directly use that:
``` python
from fastapi import APIRouter, Response, FastAPI
from fastapi.responses import ORJSONResponse

app = FastAPI()
router = APIRouter()


@router.get("/test")
async def plan_list():
    response = ORJSONResponse(list(range(10000)))
    response.headers.update({"test-header": "test-value"})
    return response


app.include_router(router)

```

**Comment by ahadrt** [comment]
@includeamin **Yes** thanks, it seems good idea.  With inspiration of your answer I found another way. what about this one?

```
from fastapi import APIRouter, Response, FastAPI
from fastapi.responses import ORJSONResponse

app = FastAPI()
router = APIRouter()


@router.get("/test")
async def plan_list(response:Response):
       response.headers["test"]="test-value"
       return ORJSONResponse(list(range(10000)),headers=response.headers)



app.include_router(router)
```

**Comment by includeamin** [comment]
> @includeamin **Yes** thanks, it seems good idea. With inspiration of your answer I found another way. what about this one?
> 
> ```
> from fastapi import APIRouter, Response, FastAPI
> from fastapi.responses import ORJSONResponse
> 
> app = FastAPI()
> router = APIRouter()
> 
> 
> @router.get("/test")
> async def plan_list(response:Response):
>        response.headers["test"]="test-value"
>        return ORJSONResponse(list(range(10000)),headers=response.headers)
> 
> 
> 
> app.include_router(router)
> ```

sounds good!

**Comment by ahadrt** [comment]
@includeamin  thanks once more . I close this issue

**Comment by tiangolo** [comment]
Thanks for the help @includeamin ! 🚀 

And thanks for closing the issue @ahadrt ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
